### DISCUSSION TITLE
Question about using impulse with object direction

### DISCUSSION THREAD

**Discussion by khoamaitan** [jrouwe/JoltPhysics#981] [Q&A] [answered]
Hello,
I'm implementing jolt for my project. I'm having a weird behavior from jolt when adding impulse to object according to its orientation. In my case, I only apply force to the object current OZ however, the object trajectory is different based on the initial orientation of object. As you can see on the first clip, when the object is aligned with axis OZ, the box go straight a head (but sometimes it does deviate from axis OZ but less occurrent ).

https://github.com/jrouwe/JoltPhysics/assets/16591163/39226141-3699-4cc2-8cab-9a56cd423a57

On the second video, when I rotate the object around axis OY 45 degree, the object couldn't go straight forward. It has this behavior all the time

https://github.com/jrouwe/JoltPhysics/assets/16591163/24f46354-ebae-4406-ac13-26ca9b0a52bd

The box is placed on the floor like the samples in the code. I can guess vaguely why it behaves like this but I wonder if anyone have the solution for this problem ?
By the way, here is part of my code to add impulse to object before calling the "Update" function of physicsSystem
![image](https://github.com/jrouwe/JoltPhysics/assets/16591163/3a652d7e-7efb-4519-ba03-c593b3e277f7)
The deltaT is constant (0.016) between loops. There is only "force Z" applied to object. Vector (px,py,pz) is zero
Thank you in advanced

**Comment by jrouwe** [comment]
This is caused by small inaccuracies in the friction calculation that build up over time if you apply impulses in local space. If you set the friction on the floor to zero you will see that it stops happening.

**Reply by khoamaitan** [reply]
Thank you for you response. I tested all day and it's true that without friction or gravity, the phenomenon doesn't happen. also ran the test  (with friction and gravity) where the applied force is constant in global space (video below). And the box does drift a lot. In the test, the force vector (fx,0,fy) is applied at the center of the box where fx = fy 

https://github.com/jrouwe/JoltPhysics/assets/16591163/4ccf7527-f491-4d2c-8adc-b759e6cbae56

It bugs me that the deviation happens more often when the object is not aligned with axes OZ or OX
May be you know why  there is such difference ?

**Reply by jrouwe** [reply, marked as answer]
![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/8526dbd6-c35f-4f7a-801a-2ca5f6e67c70)

The box has 4 contact points (marked 1 .. 4) and the friction axis are aligned with the world space axis in this case (red/green arrows, which is why it is less apparent if the box is aligned with the world space axis). The contact points are solved in order so that means when solving 1 the box starts rotating a little bit because the force is off center (but 3 should counter that rotation because it operates in the opposite direction), same for 2 and 4. Because the frictions are applied sequentially instead of all at once, they're not 100% countering each other so that causes drift. I'll check if there's a better solution for it.
